mm2src/rpc_task/src/task.rs

```diff
@@ -6,15 +6,20 @@ use serde::Serialize;
 pub trait RpcTaskTypes {
     type Item: Serialize + Clone + Send + Sync + 'static;
     type Error: SerMmErrorType + Clone + Send + Sync + 'static;
-    type InProgressStatus: Clone + Send + Sync + 'static;
-    type AwaitingStatus: Clone + Send + Sync + 'static;
+    type InProgressStatus: Serialize + Clone + Send + Sync + 'static;
+    type AwaitingStatus: Serialize + Clone + Send + Sync + 'static;
     type UserAction: NotMmError + Send + Sync + 'static;
 }
 
 #[async_trait]
 pub trait RpcTask: RpcTaskTypes + Sized + Send + 'static {
     fn initial_status(&self) -> Self::InProgressStatus;
 
+    /// Returns the ID of the client that initiated/requesting the task.
+    ///
+    /// This is related to event streaming and is used to identify the client to whom the updates shall be sent.
+    fn client_id(&self) -> Option<u64>;
```

Comment:
Task managed RPC are favored to be sent over SSE. that's why we have client id here. it's perfectly fine though for the implementor of a new `dyn RpcTask` to set this to `None` if they don't want such task to stream out the updates in real time.

Reply:
Oh, so this implements request-response functionality to SSE layer? It feels like we should switch to Websockets to cover this need.

Reply:
> Oh, so this implements request-response functionality to SSE layer?

request is through rpc, and the response data is still available via rpc but we also send updates through SSE (for free) if:
1- this rpc task implementation supports it (& why wouldn't it), and...
2- the client provided their ID in the `task::init` request (it's optional, the client might just omit it and nobody will receive these updates in real time, they have to query `task::get_status` periodically then to receive the updates), and...
3- the client is actually connected via an SSE connection, otherwise real time updates are thrown into the sea.

> It feels like we should switch to Websockets to cover this need.

i can't find it rn but we actually talked about this before and SSE was lighter somehow than websockets. im no expert in this tho :/

Reply:
It's true that SSE is lighter since it only handles one-way communication. But in our case, we are doing workarounds (sending requests through RPC and getting responses via SSE) to make it two-way. When we do that, not only that SSE becomes heavier than WebSockets (as Websockets naturally support two-way communication) our project also becomes harder to maintain with all these complexities.

There’s another option: we could extend our RPC layer with WebSocket connections (I know, this is a huge development block for this task, but IMO it's worth it.). We would use this for two things: regular request/response communication and event streaming. This would let us avoid the complexity and overhead of using SSE for two-way communication. Plus, we would have WebSockets as an extra protocol for RPCs. A lot of projects already do this like ETH and Cosmos nodes (you can set up a WebSocket connection and invoke RPC requests directly - like electrums but with Websockets).

Reply:
> When we do that, not only that SSE becomes heavier than WebSockets (as Websockets naturally support two-way communication)

but does it really get heavier? the 2 way communication we have is hugely unbalanced (server sends more bit than client, client requests are usually very short commands with no data what so ever).

> our project also becomes harder to maintain with all these complexities.

complexity wise, i think we would still need the same pub/sub arch & client_id identification (or a similar mechanism, one shot response channel for instance) to manage where responses are supposed to go to, that's just the nature of allowing multi-client arch. The streamers arch i guess would be the same as well, not that we would spawn a new streamer for every client.

Reply:
> but does it really get heavier? the 2 way communication we have is hugely unbalanced (server sends more bit than client, client requests are usually very short commands with no data what so ever).

With all these hacked RPC + SSE communications, it gets heavier, yes.

> complexity wise, i think we would still need the same pub/sub arch & client_id identification (or a similar mechanism, a response channel for instance) to manage where responses are supposed to go to, that's just the nature of allowing multi-client arch. The streamers arch i guess would be the same as well, not that we would spawn a new streamer for every client.

We don't need `client-id` stuff. I can use built-in JSON-RPC "id" field for characterizing my requests.

Reply:
> We don't need client-id stuff. I can use built-in JSON-RPC "id" field for characterizing my requests.

Discussed over DM. client ids will be made optionals defaulting to `0` for simplicity in single client case, same for initial SSE connection (defaults to `0` when no ID is given).

later we might add websockets interface for the RPC which could make us auto-generate the client IDs from within KDF without the need for clients to generate or remember it (because the client will always talk to us through the same websocket, that's how we know/identify them). https://github.com/KomodoPlatform/komodo-defi-framework/issues/2207#issue-2497161303

Reply:
done here [91d98ed](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2172/commits/91d98edbefc3edd1b68c8bbf912c2db31a5718ed)